channeldb/kvdb/etcd/commit_queue.go

```diff
@@ -0,0 +1,146 @@
+// +build kvdb_etcd
+
+package etcd
+
+import (
+	"context"
+	"sync"
+)
+
+// commitQueueSize is the maximum number of commits we let to queue up. All
+// remaining commits will block on commitQueue.Add().
+const commitQueueSize = 100
+
+// commitQueue is a simple execution queue to manage conflicts for transactions
+// and thereby reduce the number of times conflicting transactions need to be
+// retried. When a new transaction is added to the queue, we first upgrade the
+// read/write counts in the queue's own accounting to decide whether the new
+// transaction has any conflicting dependencies. If the transaction does not
+// conflict with any other, then it is comitted immediately, otherwise it'll be
+// queued up for later exection.
+// The algorithm is described in: http://www.cs.umd.edu/~abadi/papers/vll-vldb13.pdf
+type commitQueue struct {
+	ctx       context.Context
+	mx        sync.Mutex
+	readerMap map[string]int
+	writerMap map[string]int
+
+	commitMutex sync.RWMutex
+	queue       chan (func())
+	done        chan (struct{})
+}
+
+// NewCommitQueue creates a new commit queue, with the passed abort context.
+func NewCommitQueue(ctx context.Context) *commitQueue {
+	q := &commitQueue{
+		ctx:       ctx,
+		readerMap: make(map[string]int),
+		writerMap: make(map[string]int),
+		queue:     make(chan func(), commitQueueSize),
+		done:      make(chan struct{}),
+	}
+
+	// Start the queue consumer loop.
+	go q.mainLoop()
+
+	return q
+}
+
+// Wait waits for the queue to stop (after the queue context has been canceled).
+func (c *commitQueue) Wait() {
+	<-c.done
+}
+
+// Add increases lock counts and queues up tx commit closure for execution.
+// Transactions that don't have any conflicts are executed immediately by
+// "downgrading" the count mutex to allow concurrency.
+func (c *commitQueue) Add(commitLoop func(), rset readSet, wset writeSet) {
+	c.mx.Lock()
+	blocked := false
+
+	// Increase read counts and mark as blocked if any key is in the
+	// writer map.
+	for key := range rset {
+		c.readerMap[key] += 1
+		if !blocked {
+			blocked = c.writerMap[key] > 0
+		}
+	}
+
+	// Increase write counts and mark as blocked if any key is in the
+	// reader map or is already blocked for writing.
+	for key := range wset {
+		c.writerMap[key] += 1
+		if !blocked {
+			_, rsetContainsKey := rset[key]
+			blocked = (c.writerMap[key] > 1 ||
+				(c.readerMap[key] > 0 && !rsetContainsKey))
+		}
+	}
+
+	if blocked {
+		// Add the transaction to the queue if conflicts with an already
+		// queued one.
+		select {
+		case c.queue <- commitLoop:
```

Comment:
Ooh, a bit scary holding a mutex while sending on a channel. It *looks* like it is alright, but couldn't we really release it before sending?

What could happen is that the blocked transactions could be executed in a different order, but that shouldn't matter as we already have determined they will run exclusively.

Reply:
Correct, things could be re-ordered here, but that's ok as we're just trying to reduce contention on a bests-effort basis.